ts/sbd-server/src/index.ts

```diff
@@ -0,0 +1,434 @@
+
+      if (method !== 'GET') {
+        throw err('expected GET', 400);
+      }
+
+      const { pubKeyStr } = parsePubKey(url.pathname);
+
+      const ipId = env.RATE_LIMIT.idFromName(ip);
+      const ipStub = env.RATE_LIMIT.get(ipId) as DurableObjectStub<DoRateLimit>;
+      const { shouldBlock } = await ipStub.bytesReceived(
+        Date.now(),
+        pubKeyStr,
+        1,
+      );
+      if (shouldBlock) {
+        throw err(`limit`, 429);
+      }
+
+      // DO instanced by our pubKey
+      const id = env.SIGNAL.idFromName(pubKeyStr);
+      const stub = env.SIGNAL.get(id);
+
+      // just forward the full request / response
+      return await stub.fetch(request);
+    } catch (e: any) {
+      console.error('error', e.toString());
+      return new Response(JSON.stringify({ err: e.toString() }), {
+        status: e.status || 500,
+      });
+    }
+  },
+};
+
+/**
+ * "RATE_LIMIT" durable object.
+ * This is a thin wrapper around the "RateLimit" class.
+ */
+export class DoRateLimit extends DurableObject {
+  ctx: DurableObjectState;
+  env: Env;
+  rl: RateLimit;
+
+  constructor(ctx: DurableObjectState, env: Env) {
+    super(ctx, env);
+    this.ctx = ctx;
+    this.env = env;
+    this.rl = new RateLimit(LIMIT_NANOS_PER_BYTE, 16 * 16 * 1024);
+  }
+
+  async bytesReceived(
+    now: number,
+    pk: string,
+    bytes: number,
+  ): Promise<RateLimitResult> {
+    return this.rl.bytesReceived(now, pk, bytes);
+  }
+}
+
+/**
+ * "SIGNAL" durable object.
+ */
+export class DoSignal extends DurableObject {
+  ctx: DurableObjectState;
+  env: Env;
+  queue: { [index: string]: Array<Uint8Array> };
+  alarmLock: boolean;
+  curLimit: number;
+
+  constructor(ctx: DurableObjectState, env: Env) {
+    super(ctx, env);
+    this.ctx = ctx;
+    this.env = env;
+    this.queue = {};
+    this.alarmLock = false;
+    this.curLimit = 0;
+  }
+
+  /**
+   * Client websockets are connected to a durable object identified
+   * by their own pubKey. When they send forward messages, those must
+   * be sent to durable objects identified by the destination pubKey.
+   * This is the api for those messages to be forwarded.
+   */
+  async forward(messageList: Array<Uint8Array>) {
+    for (const ws of this.ctx.getWebSockets()) {
+      for (const message of messageList) {
+        ws.send(message);
+      }
+    }
+  }
+
+  /**
+   * This is the http endpoint for the "SIGNAL" durable object.
+   * The worker http fetch endpoint above forwards the request here.
+   * This function performs some checks, then upgrades the connection
+   * to a websocket.
+   */
+  async fetch(request: Request): Promise<Response> {
+    let cleanServer = null;
+    try {
+      const ip = request.headers.get('cf-connecting-ip') || 'no-ip';
+      const url = new URL(request.url);
+
+      const { pubKeyStr, pubKeyBytes } = parsePubKey(url.pathname);
+
+      if (this.ctx.getWebSockets().length > 0) {
+        throw err('websocket already connected', 400);
```

Comment:
Errors thrown from inside this try..catch block will all be caught by the `catch` and return the same status code.

Reply:
The second param of this `err` function sets a `.status` property on the error that is thrown (in this case `400`) the `catch` block uses this value, but defaults to `500` in case an error without a status property is thrown.

Reply:
Sorry, I didn't look closely here.